mainmatter.md

```diff
@@ -326,17 +328,19 @@ following.
 ## DPoP Proof Replay {#Token_Replay}
 
 If an adversary is able to get hold of a DPoP proof JWT, the adversary
-could replay that token later at the same endpoint (the HTTP endpoint
+could replay that token at the same endpoint (the HTTP endpoint
 and method are enforced via the respective claims in the JWTs). To
 prevent this, servers MUST only accept DPoP proofs for a limited time
 window after their `iat` time, preferably only for a brief period.
-Furthermore, the `jti` claim in each JWT MUST contain a unique
-(incrementing or randomly chosen) value, as proposed in [@!RFC7253].
-Resource servers SHOULD store values at least for the time window in
-which the respective JWT is accepted and decline HTTP requests by
-clients if a `jti` value has been seen before.
-
-Note: To acommodate for clock offsets, the server MAY accept DPoP
+Furthermore, the `jti` claim in each DPoP proof JWT MUST contain a globally unique
+value (e.g., 128 bits of pseudorandom data base64url encoded). 
```

Comment:
I think we should move the recommendation to line 107ff where we explain `jti` the first time (or a separate paragraph immediately after the enumeration there). I would suggest `... SHOULD contain 128 bits of pseudorandom data in base64url encoding (or any other suitable encoding), or a version 4 UUID according to [@RFC4122]`.

Note that UUID v4 only contains 121 or 122 bits of entropy. Should still be fine for this purpose.

Reply:
Okay, I'll update accordingly. 

I think 128 (and even 121/122) bits is probably more than is actually necessary for this. 340 undecillion is an almost incomprehensibly big number.  But I don't know what the right floor is or how to calculate it. Maybe a doctor could figure that out? :) 120? 112? maybe 96? I dunno. 128 seems like a safe choice in the absence of knowing better. 

I'm no fan of UUIDs in JWTs because the hex and superfluous dashes but it's an easy option in some cases so mentioning it seems ok.

Reply:
done in 37a351e29e0c6966bec5d2eb4c720eb9494d75b4

Reply:
If I'm not mistaken (my math is a bit rusty here), the probability of a nonce collision is less than 2^-32 if 2^17 nonces with 64 bits are chosen randomly and uniformly. Or put differently, we would expect a duplicate nonce about every 1,000,000,000 invocations of the algorithm. If we had 10000 clients, they would _each_ need to produce about 100,000 DPoP proofs _within the  average lifetime of the proofs_ for a collision. With a lifetime of one minute, this would mean that they need to create about 1600 requests per second. 

I would say, 64 bits is a safe choice. If that is not sufficient, the server can just track the tuple (access token, jti) instead of just the jti. Then **one** client would need to create 16,000,000 requests per second for a collision.

Reply:
Based on the potential issues around tracking the access token itself that we uncovered discussing "option 3" in #47 I think it'd be best to track the jti alone. 

I'm still thinking about the math... :)

Reply:
Trying to dumb down the math for myself I working primarily with "√n is roughly the number you need to have a 50% chance of a match with n items" from https://betterexplained.com/articles/understanding-the-birthday-paradox which lines up roughly with the "p = 0.50 & 0.75" columns at https://en.wikipedia.org/wiki/Birthday_problem#Probability_table and also isn't out of line with your calculations (I think anyway it'd suggest to expect a duplicate nonce about every 4,300,000,000 invocations and that's in the same order of magnitude).

Based on the square root approximation, 10000 clients and 1 minute dpop life would need each client to do around 7,200 requests/sec. Which is an awful lot.

But what about a million clients? 72 req/sec. Or more? 10 million and 7 req/sec. That's a heck of a lot of clients but internet scale and all that suggests it's not an all together unreasonable possibility. So while 64 bits probably is actually enough, I feel like it's a bit too close for comfort.

With 128 bits, a billion clients would need to be making over 300 million req/sec. That seems more than sufficient.

With 96 bits, the threshold is like a billion clients doing 4,600 req/sec. That also seems sufficient.

So... we could be conservative and stick with 128. Or I think 96 would be good too. What do you think about those options? I kinda like 96...

Reply:
I went ahead with suggesting 96 bits in 4fd069d